app/javascript/youtube.js

```diff
@@ -0,0 +1,50 @@
+const async = require('async')
+
+window.onYouTubeIframeAPIReady = () => {
+  return true
+}
+
+function setupYouTubeVideos() {
+  Array.from(document.getElementsByClassName('youtube-player')).forEach(el => {
+    if (el.dataset.loaded === '1') {
+      return
+    }
+    el.dataset.loaded = '1'
+
+    const player = new YT.Player(el.id, {
+      height: '100%',
+      width: '100%',
+      videoId: el.dataset.video_id
+    });
+
+    document.addEventListener('click', event => {
+      const segment = event.target.closest('.split');
+      if (segment === null) {
+        return
+      }
+
+      player.seekTo(parseInt(segment.dataset.start_ms / 1000 + 10))
+      setTimeout(function() { player.playVideo() }, 10) // Running this instantly after seek causes the player to hang
+    });
+
+    const ticker = document.getElementById(`video-progress-line-${el.dataset.run_id}`)
+    async.forever(next => {
+      try {
+        ticker.style.margin = `0 0 0 ${Math.max(player.getCurrentTime() - 10, 0) / (gon.scale_to / 1000) * 100}%`
+      } catch {}
+      setTimeout(() => next(), 100)
+    })
+  })
+}
+
+function waitForYouTubeAPI() {
+  if (window.YT !== undefined && typeof(window.YT.Player) === "function") {
+    setupYouTubeVideos()
+    return true
+  }
+  setTimeout(waitForYouTubeAPI, 100)
+}
+
+document.addEventListener('turbolinks:load', function() {
+  waitForYouTubeAPI()
+})
```

Comment:
The flow of code in this file feels a bit hard to follow, especially with the polling for the YouTube JS being ready. Is there a reason it needs to be a different format from what's in `twitch.js`? (i.e. a single `turbolinks:load` listener)

Reply:
The reason for the polling listener triggered by `turbolinks:load` is it was possible to hit `YT.Player(...)` before that object was fully initialized and a JS error of `YT.Player is not a constructor` would get thrown. The YouTube API JS that is included in the `<head>` is just a shim that then loads the actual API JS code asynchronously, so lots of load-order shenanigans are possible.

I'd like to get rid of the complexity and what amounts to basically busy waiting and am open to suggestions.

Reply:
In that case, is it possible the [code][1] in `twitch.js` encounters the same problem?

Also I think there's an `onload="doAThing()"` attribute you can add to `<script>` tags, maybe that would help with getting rid of the polling at least.

[1]: https://github.com/glacials/splits-io/blob/master/app/javascript/twitch.js#L10

Reply:
No, the Twitch code is fine. The script that loads is the complete JS.

The `onload` attribute probably wouldn't help since we need the script that YouTube downloads after the initial script loads.

Reply:
Ahh I see, the script we include downloads another script! Sounds crazy and awkward to deal with.

Their [iframe player API reference][1] seems to say the script tries to call a global function called `onYouTubeIframeAPIReady` when it's loaded, can we take advantage of that?

[1]: https://developers.google.com/youtube/iframe_api_reference#Getting_Started

Reply:
I think it won't play nicely with Turbolinks. On the initial page load, that function would probably get called after the DOM finished loading, but on subsequent partial loads, I'm don't think it would even get called since the JS wouldn't get reloaded. Even if it did, I don't think you can rely on the order of operations due to the async nature of the API JS.